cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java

```diff
@@ -136,6 +148,56 @@ public static void validateConfig(SecureStore secureStore, SourceControlConfig s
     }
   }
 
+  /**
+   * Commits and pushes all changes under the repository root path
+   *
+   * @param commitMeta Details for the commit including author, committer and commit message
+   * @param filesChanged List of relative path to repository root where the files are updated
+   * @throws UnexpectedRepositoryChangesException when there are changes outside the repository base path
+   * @throws GitAPIException                      when the underlying git commands fail
+   * @throws NoChangesToPushException             when there's no file changes for the commit
+   * @return the pushed commit ID
+   */
+  public String commitAndPush(CommitMeta commitMeta, List<Path> filesChanged)
+    throws UnexpectedRepositoryChangesException, NoChangesToPushException, GitAPIException {
+    validateInitialized();
+
+    // if the status is clean skip
+    Status preStageStatus = git.status().call();
+    if (preStageStatus.isClean()) {
+      throw new NoChangesToPushException("No changes have been maid for the applications to push.");
+    }
+
+    for (Path path : filesChanged) {
+      git.add().addFilepattern(path.toString()).call();
+    }
+
+    RevCommit commit = getCommitCommand(commitMeta).call();
+
+    // Fail if the status is not clean i.e. there are unexpected changes.
+    Status postCommitStatus = git.status().call();
+    if (!postCommitStatus.isClean()) {
+      throw new UnexpectedRepositoryChangesException(String.format("Unexpected changes other than changed files %s",
+                                                                   filesChanged));
+    }
+
+    PushCommand pushCommand = createCommand(git::push, sourceControlConfig, credentialsProvider);
+    pushCommand.call();
```

Comment:
does this return useful information (like the fileHash of whatever was committed)? 

It would be much better to return that information from this call rather than having a separate method to get the hash.

Reply:
yes but it is doing silly things like checking out the commit then restoring whatever it was previously. It is much cleaner to get all the information once, I get the feeling that the getFileHash() method shouldn't exist.

Reply:
Return the fileHash now, instead of calling `getFileHash()`

Reply:
do we need to check anything in the return value of this call() method to make sure it was successful? Or do errors always result in an exception thrown?

Reply:
Added checking foe the returned `PushResult`